presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +582,193 @@ private static NestedField nestedField(String name)
             private final StandardFunctionResolution functionResolution;
             private final ExpressionOptimizer expressionOptimizer;
             private final ConnectorSession connectorSession;
-
-            private SubfieldExtractor(StandardFunctionResolution functionResolution, ExpressionOptimizer expressionOptimizer, ConnectorSession connectorSession)
+            private final FunctionAndTypeManager functionAndTypeManager;
+            private final boolean isPushDownSubfieldsFromLambdasEnabled;
+
+            private SubfieldExtractor(
+                    StandardFunctionResolution functionResolution,
+                    ExpressionOptimizer expressionOptimizer,
+                    ConnectorSession connectorSession,
+                    FunctionAndTypeManager functionAndTypeManager,
+                    boolean isPushDownSubfieldsFromLambdasEnabled)
             {
                 this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
                 this.expressionOptimizer = requireNonNull(expressionOptimizer, "expressionOptimizer is null");
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
```

Comment:
why do we still need special casing for subscript and element_at functions? Why couldn't they be described by the function descriptor.

Reply:
ELEMENT_AT is a special function. FunctionDescriptor is not designed to handle it.
essentially, this allows to extract subfields when user `element_at` function instead of subscript
`map_or_array_column[1]` vs `element_at(map_or_array_column, 1)`. semantically they are the same but syntactically, they are different. 
The only thing that I can do here is to move this check inside the [FunctionResolution](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/relational/FunctionResolution.java#L251-L254). However, I do not think that I should. In different context, these function could be different.